src/OrchardCore.Modules/OrchardCore.AutoSetup/AutoSetupMiddleware.cs

```diff
@@ -0,0 +1,182 @@
+            var setupService = httpContext.RequestServices.GetRequiredService<ISetupService>();
+
+            _logger.LogInformation("AutoSetup is initializing the site");
+            if (await SetupTenant(setupService, _setupOptions.Value.RootTenant, currentShellSettings))
+            {
+                foreach (var tenant in _setupOptions.Value.SubTenants)
+                {
+                    if(!await SetupTenant(setupService, tenant, CreateTenantSettings(tenant)))
+                    {
+                        break;
+                    }
+                }
+
+                httpContext.Response.Redirect("/");
+            }
+
+            await _next.Invoke(httpContext);
+        }
+
+        /// <summary>
+        /// Setup tenant.
+        /// </summary>
+        /// <param name="setupService"> The setup service. </param>
+        /// <param name="setupOptions"> The tenant setup options. </param>
+        /// <param name="shellSettings"> The tenant shell settings. </param>
+        /// <returns>
+        /// The <see cref="Task"/>.
+        /// </returns>
+        public async Task<bool> SetupTenant(ISetupService setupService, BaseTenantSetupOptions setupOptions, ShellSettings shellSettings)
+        {
+            var setupContext = await GetSetupContext(setupOptions, setupService, shellSettings);
+
+            await setupService.SetupAsync(setupContext);
+
+            if (setupContext.Errors.Count == 0)
+            {
+                _logger.LogInformation($"AutoSetup successfully provisioned the site {setupOptions.SiteName}");
+
+                return true;
+            }
+
+            var stringBuilder = new StringBuilder();
+            foreach (var error in setupContext.Errors)
+            {
+                stringBuilder.Append($"{error.Key} : '{error.Value}'");
+            }
+
+            _logger.LogError($"AutoSetup failed installing the site '{setupOptions.SiteName}' with errors: {stringBuilder}");
+            return false;
+        }
+
+        /// <summary>
+        /// Create tenant shell settings.
+        /// </summary>
+        /// <param name="setupOptions"> The setup options. </param>
+        /// <returns> The <see cref="ShellSettings"/>. </returns>
+        public ShellSettings CreateTenantSettings(TenantSetupOptions setupOptions)
+        {
+            var shellSettings = _shellSettingsManager.CreateDefaultSettings();
+
+            shellSettings.Name = setupOptions.SiteName;
+            shellSettings.RequestUrlHost = setupOptions.RequestUrlHost;
+            shellSettings.RequestUrlPrefix = setupOptions.RequestUrlPrefix;
+            shellSettings.State = TenantState.Uninitialized;
+
+            shellSettings["ConnectionString"] = setupOptions.DatabaseConnectionString;
+            shellSettings["TablePrefix"] = setupOptions.DatabaseTablePrefix;
+            shellSettings["DatabaseProvider"] = setupOptions.DatabaseProvider;
+            shellSettings["Secret"] = Guid.NewGuid().ToString();
```

Comment:
What is that?

Reply:
I have to create a tenant shell before I setup one. (I do not have to do it for a Default tenant since shell gets created on the first start.)
I saw it in the `OrchardCore.Tenants.ApiController.Create` Action. I think for better reusability we should create an abstraction to manage tenants, but not in this PR

Reply:
>shellSettings["Secret"] = Guid.NewGuid().ToString();

I do not know, it is from OrchardCore.Tenants.ApiController.Create

Commit d918b13c
Adding configuration support (#2824)

Reply:
Yes, i think that @sebastienros was only talking about the `Secret`, it is used to create an encrypted token that is appended to the url to be able to setup a tenant that was created and still in the `Uninitialized` state, e.g. through the admin tenant page, the secret is generated when we create a tenant, and then re-used and encrypted for security to provide the admin link to setup the tenant.

So here that's okay to generate a secret, as in the api controller, that you will not use as you create the tenant and then setup it yourself, bypassing any setup controller, so you don't have to generate a protected url using a dataProtector.

So the secret is used between the creation and the setup of a given tenant, so here for me that's okay, unless we would want a security token for the url that will trigger an auto setup = create + setup tenant(s) ;)